x/smart/dns.go

```diff
@@ -0,0 +1,217 @@
+// Copyright 2024 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package smart
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"net"
+	"time"
+	"unicode"
+
+	"github.com/Jigsaw-Code/outline-sdk/dns"
+	"golang.org/x/net/dns/dnsmessage"
+)
+
+// makeFullyQualified makes the domain fully-qualified, ending on a dot (".").
+// This is useful in domain resolution to avoid ambiguity with local domains
+// and domain search.
+func makeFullyQualified(domain string) string {
```

Comment:
Shall we take care of `localhost` here? Cuz we should not append `.` to `localhost`?